File: python/euler/bin/p0083.py
# project euler: problem 83

#   I used Dijkstra's algorithm to solve.
#     https://en.wikipedia.org/wiki/Dijkstra%27s_algorithm

import sys
from typing import IO

from euler.lib.util import HeapQueue


def make_neighbor_tbl(rows: int, columns: int) -> list[list[list[tuple[int, int]]]]:
    tbl: list[list[list[tuple[int, int]]]] = [[[(0, 0)]] * columns for _ in range(rows)]
    for r in range(rows):
        for c in range(columns):
            tbl[r][c] = list(
                filter(
                    lambda tpl: tpl[0] >= 0 and tpl[0] < rows and tpl[1] >= 0 and tpl[1] < columns,
                    [(r - 1, c), (r + 1, c), (r, c - 1), (r, c + 1)],
                )
            )

    return tbl


def make_distance_tbl(rows: int, columns: int) -> list[list[int]]:
    return [[sys.maxsize] * columns for _ in range(rows)]


def compute(fh: IO) -> str:
    def parse_data(fh: IO) -> list[list[int]]:
        return [list(map(int, line.split(','))) for line in fh.read().splitlines()]

    matrix = parse_data(fh)
    nbr_tbl = make_neighbor_tbl(rows=len(matrix), columns=len(matrix[0]))
    dist_tbl = make_distance_tbl(rows=len(matrix), columns=len(matrix[0]))
    dist_tbl[0][0] = matrix[0][0]
    pq = HeapQueue()
    pq.insert((dist_tbl[0][0], (0, 0)))

    while not pq.is_empty():
        d, (i, j) = pq.extract()
        for x, y in nbr_tbl[i][j]:
            if (new_d := d + matrix[x][y]) < dist_tbl[x][y]:
                dist_tbl[x][y] = new_d
                pq.insert((new_d, (x, y)))

    return str(dist_tbl[-1][-1])


def solve() -> str:
    from euler.lib.resource import asset_file

    fh = asset_file('p083_matrix.txt')
    result = compute(fh)
    fh.close()
    return result

File: typescript/euler/bin/p0083.ts
// project euler: problem 83

import { ascend, BinaryHeap } from "std/data_structures/mod.ts";
import { assetData } from "../lib/asset.ts";
import { range } from "../lib/util.ts";

const makeNeighborTbl = (
  rows: number,
  cols: number,
): [number, number][][][] => {
  let tbl: [number, number][][][] = [];
  for (const _ of range(0, rows)) {
    tbl = tbl.concat([new Array(cols).fill([])]);
  }

  for (const r of range(0, rows)) {
    for (const c of range(0, cols)) {
      //console.log(  [[r - 1, c], [r + 1, c], [r, c - 1], [r, c + 1]].filter((x) => x[0] >= 0 && x[0] < rows && x[1] >= 0 && x[1] < cols) as [number, number][] );
      tbl[r][c] = [[r - 1, c], [r + 1, c], [r, c - 1], [r, c + 1]].filter((x) =>
        x[0] >= 0 && x[0] < rows && x[1] >= 0 && x[1] < cols
      ) as [number, number][];
    }
  }
  return tbl;
};

const makeDistanceTbl = (rows: number, cols: number): number[][] => {
  let tbl: number[][] = [];
  for (const _ of range(0, rows)) {
    tbl = tbl.concat([new Array(cols).fill(Number.MAX_SAFE_INTEGER)]);
  }

  return tbl;
};

export const compute = (data: string): string => {
  const parseData = (data: string): number[][] => {
    const splitLines = (str: string): string[] => {
      const result = str.split(/\r?\n/);
      if (result.at(-1) === "") {
        return result.slice(0, -1);
      } else {
        return result;
      }
    };

    return splitLines(data).map((x) => x.split(",").map((y) => Number(y)));
  };

  const matrix = parseData(data);
  const nbrTbl = makeNeighborTbl(matrix.length, matrix[0].length);
  const distTbl = makeDistanceTbl(matrix.length, matrix[0].length);
  distTbl[0][0] = matrix[0][0];

  const pq = new BinaryHeap<[number, [number, number]]>((a, b) =>
    ascend(a[0], b[0])
  );
  pq.push([distTbl[0][0], [0, 0]]);

  while (pq.isEmpty() === false) {
    const [d, [i, j]] = pq.pop() as [number, [number, number]];
    for (const [x, y] of nbrTbl[i][j]) {
      const new_d = d + matrix[x][y];
      if (new_d < distTbl[x][y]) {
        distTbl[x][y] = new_d;
        pq.push([new_d, [x, y]]);
      }
    }
  }

  return String(distTbl.at(-1)!.at(-1));
};

export const solve = (): string => {
  const data = new TextDecoder().decode(assetData("p083_matrix.txt"));
  return compute(data);
};
